build/tasks/CodeGen/GenerateSubmoduleGraph.cs

```diff
@@ -0,0 +1,228 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.IO;
+using System.Text;
+using System.Threading;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using NuGet.Frameworks;
+using NuGet.Versioning;
+using RepoTools.BuildGraph;
+using RepoTasks.ProjectModel;
+using RepoTasks.Utilities;
+using RepoTasks.CodeGen;
+using NuGet.Packaging.Core;
+
+namespace RepoTasks
+{
+    public class GenerateSubmoduleGraph : Task, ICancelableTask
+    {
+        private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+
+        /// <summary>
+        /// Repositories that we are building new versions of.
+        /// </summary>
+        [Required]
+        public ITaskItem[] Solutions { get; set; }
+
+        [Required]
+        public ITaskItem[] Artifacts { get; set; }
+
+        [Required]
+        public ITaskItem[] Repositories { get; set; }
+
+        [Required]
+        public string RepositoryRoot { get; set; }
+
+        [Required]
+        public string Properties { get; set; }
+
+        public void Cancel()
+        {
+            _cts.Cancel();
+        }
+
+        public override bool Execute()
+        {
+            var packageArtifacts = Artifacts.Select(ArtifactInfo.Parse)
+                .OfType<ArtifactInfo.Package>()
+                .Where(p => !p.IsSymbolsArtifact)
+                .ToDictionary(p => p.PackageInfo.Id, p => p, StringComparer.OrdinalIgnoreCase);
+
+            var factory = new SolutionInfoFactory(Log, BuildEngine5);
+            var props = MSBuildListSplitter.GetNamedProperties(Properties);
+
+            Log.LogMessage(MessageImportance.High, $"Beginning cross-repo analysis on {Solutions.Length} solutions. Hang tight...");
+
+            if (!props.TryGetValue("Configuration", out var defaultConfig))
+            {
+                defaultConfig = "Debug";
+            }
+
+            var solutions = factory.Create(Solutions, props, defaultConfig, _cts.Token).OrderBy(f => f.Directory).ToList();
+            Log.LogMessage($"Found {solutions.Count} and {solutions.Sum(p => p.Projects.Count)} projects");
+
+            if (_cts.IsCancellationRequested)
+            {
+                return false;
+            }
+
+            return GenerateGraph(packageArtifacts, solutions);
+        }
+
+        private bool GenerateGraph(IDictionary<string, ArtifactInfo.Package> packageArtifacts, IReadOnlyList<SolutionInfo> solutions)
+        {
+            var repoGraph = new AdjacencyMatrix(solutions.Count);
+            var packageToProjectMap = new Dictionary<PackageIdentity, ProjectInfo>();
+
+            for (var i = 0; i < solutions.Count; i++)
+            {
+                var sln = repoGraph[i] = solutions[i];
+
+                foreach (var proj in sln.Projects)
+                {
+                    if (!proj.IsPackable
+                        || proj.FullPath.Contains("samples")
+                        || proj.FullPath.Contains("tools/VS.Web.CG.Design"))
```

Comment:
Out of curiosity, what's special about this?

Reply:
There are multiple csproj files in this folder that produce the same package ID, just for different platforms (x64/arm/x86). We should probably just fix scaffolding's code.